### PR TITLE
Implement changes for dynamical offsets

### PR DESCRIPTION
**Pull request by taldcroft**
Implement changes to support checking PCAD attitudes for loads created with Matlab tools 2016_210 and later. This implements the Cycle 18 aimpoint transition plan with ACA CCD-based dynamical aimpoint offsets.

Goes along with https://github.com/sot/hopper/pull/8.

### CODE REVIEW

**Comment by jeanconn** [comment]
And in the vein of testing, should we add one of the dynamic test products directories to the starcheck regression directories?

**Comment by jeanconn** [comment]
And of note, we haven't really defined a process to run the regression tests when they depend on changes in other directories, so for this case I've done:

```
-export PYTHONPATH=$DIR:$PYTHONPATH
+export PYTHONPATH=$DIR:/home/jeanconn/git/hopper
```

in sandbox_starcheck.  I didn't want to just set PYTHONPATH to my hopper directory in general because I assume it would get executed on all of the baseline runs of starcheck (we still run this as a run-the-old starcheck, then run the new starcheck in the regression outputs).

**Comment by taldcroft** [comment]
Here is the way I have been "unit-testing" this:
```
import starcheck.pcad_att_check

starcheck.pcad_att_check.make_pcad_attitude_check_report(
    '/proj/sot/ska/ops/SFE/MAR0516M/ofls/CR065_0406.backstop',
    '/proj/sot/ska/ops/SFE/MAR0516M/ofls/mps/or/MAR0516_A.or',
    '/proj/sot/ska/ops/SFE/MAR0516M/ofls/mps/mm065_0409.sum',
    '/proj/sot/ska/ops/SFE/MAR0516M/ofls/History/SIMTRANS.txt',
    '/proj/sot/ska/ops/SFE/MAR0516M/ofls/History/SIMFOCUS.txt',
    '/proj/sot/ska/ops/SFE/MAR0516M/ofls/mps/ode/characteristics/CHARACTERIS_28FEB16',
    'MAR0516M_att_check.txt',
    '/proj/sot/ska/ops/SFE/MAR0516M/ofls/output/MAR0516M_dynamical_offsets.txt')

starcheck.pcad_att_check.make_pcad_attitude_check_report(
    '/proj/sot/ska/ops/SFE/MAR0516O/ofls/CR065_0405.backstop',
    '/proj/sot/ska/ops/SFE/MAR0516O/ofls/mps/or/MAR0516_A.or',
    '/proj/sot/ska/ops/SFE/MAR0516O/ofls/mps/mm065_0409.sum',
    '/proj/sot/ska/ops/SFE/MAR0516O/ofls/History/SIMTRANS.txt',
    '/proj/sot/ska/ops/SFE/MAR0516O/ofls/History/SIMFOCUS.txt',
    '/proj/sot/ska/ops/SFE/MAR0516O/ofls/mps/ode/characteristics/CHARACTERIS_28FEB16',
    'MAR0516O_att_check.txt',
    '/proj/sot/ska/ops/SFE/MAR0516O/ofls/output/MAR0516O_dynamical_offsets.txt')

starcheck.pcad_att_check.make_pcad_attitude_check_report(
    '/proj/sot/ska/ops/SFE/JUL0415M/ofls/CR185_1017.backstop',
    '/proj/sot/ska/ops/SFE/JUL0415M/ofls/mps/or/JUL0415_A.or',
    '/proj/sot/ska/ops/SFE/JUL0415M/ofls/mps/mm185_1000.sum',
    '/proj/sot/ska/ops/SFE/JUL0415M/ofls/History/SIMTRANS.txt',
    '/proj/sot/ska/ops/SFE/JUL0415M/ofls/History/SIMFOCUS.txt',
    None,
    'JUL0415M_att_check.txt',
    '/proj/sot/ska/ops/SFE/JUL0415M/ofls/output/JUL0415M_dynamical_offsets.txt')

starcheck.pcad_att_check.make_pcad_attitude_check_report(
    '/proj/sot/ska/ops/SFE/JUL0415O/ofls/CR185_1018.backstop',
    '/proj/sot/ska/ops/SFE/JUL0415O/ofls/mps/or/JUL0415_A.or',
    '/proj/sot/ska/ops/SFE/JUL0415O/ofls/mps/mm185_1000.sum',
    '/proj/sot/ska/ops/SFE/JUL0415O/ofls/History/SIMTRANS.txt',
    '/proj/sot/ska/ops/SFE/JUL0415O/ofls/History/SIMFOCUS.txt',
    None,
    'JUL0415O_att_check.txt',
    '/proj/sot/ska/ops/SFE/JUL0415O/ofls/output/JUL0415O_dynamical_offsets.txt')

```

**Comment by taldcroft** [comment]
About the test regression directories, let's defer that for a separate discussion of "what is the function of regression test directories".  In this case we clearly have a functional change which is being validated with a functional test, albeit not one that will be regularly repeated.  I don't fully understand the process but don't have time to have that discussion now.

**Comment by taldcroft** [comment]
About the regression test and sandboxing, again I don't fully understand the process but I thought that the "baseline" should always be running the installed flight starcheck.  In that case you can safely unset `PYTHONPATH` for that bit since you never need that configurability for a run of flight starcheck.

**Comment by jeanconn** [comment]
Right, that's what I'm saying.   But since the flight starcheck and the test starcheck both run from the same "run_regress" script, I can't just set the PYTHONPATH in my shell before I call that script.  Instead, I need to set it just for the test starcheck.  I could either do that by changing the run_regress script to set PYTHONPATH as it calls `sandbox_starcheck` or I could, as I have, just add the PYTHONPATH I want to `sandbox_starcheck` for the run of the tool.

**Comment by jeanconn** [comment]
I was writing down what I did here more so there'd be a record than for much input on that, but input certainly welcome.  I know you want to scrap the whole system, but I'm not sure how much to try to work with what we have for validation.  

For the  "separate discussion of ""what is the function of regression test directories""", we've previously used that to have these regression outputs whenever we've had new types of activities to test against and / or new checks from starcheck.  It seems like these changes in response to the new process are both new behaviors in products and new checks in starcheck, but if we are moving away from the regression list into more functional testing, that makes sense too.  Either way, this is no hold-up on my end.

**Comment by taldcroft** [comment]
I just thought that for the line in `run_regress` that actually calls flight starcheck you can use `env PYTHONPATH= ...` (along with setting PYTHONPATH as part of the test process in this case) and it'll be fine.  That should generally address this problem with configured code / process instead of always requiring a hand-edit.  Or am I not understanding?

**Comment by taldcroft** [comment]
Thanks for feedback on the regression file process.  As mentioned I don't have time for the discussion now, we'll take it up later.

**Comment by jeanconn** [comment]
You want me to edit run_regress to call flight starcheck by explicitly unsetting PYTHONPATH in the call?  OK.  That works

**Comment by taldcroft** [comment]
> You want me to edit run_regress to call flight starcheck by explicitly unsetting PYTHONPATH in the call?

Yes assuming (as you imply) that it works.

**Comment by jeanconn** [comment]
Actually, it turns out this isn't necessary, so I'm going to roll it back.  Ugh.

**Comment by jeanconn** [comment]
And by not-necessary, I meant that /proj/sot/ska/bin/starcheck already handles this for us with:
```
unset PYTHONPATH
```
I had not tested many PYTHONPATH changes up to thispoint.

**Comment by jeanconn** [comment]
35cb4cc will add this output if the dynamic aimpoint file is missing in products that start after 21-Aug-2016 (output tested by setting the date in in the past).
```
------------  VERIFY ATTITUDE (SI_ALIGN CHECK)  -----------------

Error.  dynamic aimpoint file not found. 
------------  CCD TEMPERATURE PREDICTION -----------------
```

File: src/starcheck.pl [hunk 1]
```diff
@@ -691,11 +696,16 @@ sub json_obsids{
     if (not defined $char_file){
         $out .= "Error.  Characteristics file not found. \n";
     }
+    # dynamic aimpoint files are required after 21-Aug-2016
+    my $AIMPOINT_REQUIRED_AFTER = '2016:234:00:00:00.000';
+    if ((not defined $aimpoint_file) and ($bs[0]->{time} > date2time($AIMPOINT_REQUIRED_AFTER))){
+        $out .= "Error.  dynamic aimpoint file not found. \n";
+    }
```
Comment (jeanconn) [inline src/starcheck.pl]: Whoops, I just broke this if-else.  The aimpoint file check should be an independent check before the if-else on the characteristics file.

**Comment by jeanconn** [comment]
Error output will look like this (tested by setting the dynamic file required date to a date in the past).
```
------------  VERIFY ATTITUDE (SI_ALIGN CHECK)  -----------------

Error.  dynamic aimpoint file not found. 
[OK] Coordinates as expected.
[OK] Characteristics file newer than 30 days

------------  CCD TEMPERATURE PREDICTION -----------------
```

**Comment by taldcroft** [comment]
:+1:

**Comment by jeanconn** [comment]
Closes #157

File: src/starcheck.pl [hunk 2]
```diff
@@ -688,14 +693,22 @@ sub json_obsids{
 my $CHAR_REQUIRED_AFTER = '2015:315:00:00:00.000';
 if ((defined $char_file) or ($bs[0]->{time} > date2time($CHAR_REQUIRED_AFTER))){
     $out .= "------------  VERIFY ATTITUDE (SI_ALIGN CHECK)  -----------------\n\n";
+    # dynamic aimpoint files are required after 21-Aug-2016
+    my $AIMPOINT_REQUIRED_AFTER = '2016:234:00:00:00.000';
+    if ((not defined $aimpoint_file) and ($bs[0]->{time} > date2time($AIMPOINT_REQUIRED_AFTER))){
+        $out .= "Error.  dynamic aimpoint file not found. \n";
+    }
+    # The attitude checks are not possible without the char file but are possible without the
```
Comment (taldcroft) [inline src/starcheck.pl]: I should have seen this comment in the diffs and noted that the first clause is incorrect.  My bad.

Comment (jeanconn) [inline src/starcheck.pl]: It is well-commented as far as bugs go!

**Pull request merged**